File: programas/metodo-biseccion/biseccion.go
package main

import (
	"fmt"
	"math"
)

func main() {
	n := 13.0
	tol := 0.5 * math.Pow(10, 2 - n)
	root, eps_a := bisect(f, 0, 10, tol, 0)
	fmt.Println(root, "\t", eps_a)
}

func bisect(f func(x float64) float64, low, up, tol, depth float64) (float64, float64) {
	middle := (low + up) / 2.0

	eps_a := math.Abs((low - middle) / middle) * 100

	if depth > 100 || eps_a < tol {
		return middle, eps_a
	}

	if f(low) * f(middle) < 0 {
		return bisect(f, low, middle, tol, depth + 1)
	} else {
		return bisect(f, middle, up, tol, depth + 1)
	}
}

func f(x float64) float64 {
	return x*x - 4 - math.Sin(x)
}

File: programas/metodo-biseccion/biseccion.f90
PROGRAM biseccion
        IMPLICIT NONE
        REAL (KIND=16) root, bisect, eps_s
        INTEGER n

        n = 14
        eps_s = 0.5d0 * (10.d0 ** (2 - n))
        root = bisect(-0._16, 5._16, eps_s, 0)
END PROGRAM biseccion

FUNCTION f(x)
        IMPLICIT NONE
        REAL (KIND=16) f, x
        f = (x**2) - 4.d0 - SIN(x)
END FUNCTION

RECURSIVE FUNCTION bisect(xl, xu, eps_s, depth) RESULT(res)
        IMPLICIT NONE
        REAL (KIND=16) xl, xu, xr, res, prev, f, eps_s, eps_a
        INTEGER depth

        xr = (xl + xu) / 2.d0

        eps_a = ABS((xr - xl) / xr) * 100
        IF (depth > 100 .or. eps_a < eps_s) THEN
                res = xr
                WRITE(*,*) 'root:', res, 'eps_a', eps_a
                RETURN
        END IF

        IF (f(xl) * f(xr) < 0) THEN
                res = bisect(xl, xr, eps_s, depth + 1)
        ELSE
                res = bisect(xr, xu, eps_s, depth + 1)
        END IF
        res = 0
END FUNCTION
